### DISCUSSION TITLE
ModelValidation with separate model class

### DISCUSSION THREAD

**Discussion by Catalan-1ap** [canton7/Stylet#336] [Q&A] [answered]
ModelValidation does not work with the property of the object being validated. 

For example, there is a ViewModel:
```csharp

public sealed class PublicationViewModel : Screen
{
	public Publication? Result { get; private set; }
	public string? Title { get; private set; }

	public bool CanAccept => !HasErrors;


	public PublicationViewModel()
	{
		Result = new()
		{
			Address      = new(),
			Subscription = new()
		};

		Validator = new FluentValidationAdapter<PublicationViewModel>(new PublicationViewModelValidator());
		Validate();
	}


	public void Accept()
	{
		if (Validate())
			RequestClose(true);
	}


	protected override void OnValidationStateChanged(IEnumerable<string> changedProperties)
	{
		base.OnValidationStateChanged(changedProperties);

		NotifyOfPropertyChange(() => CanAccept);
	}
}
```

Validator for it:
```csharp

public sealed class PublicationViewModelValidator : AbstractValidator<PublicationViewModel>
{
	public PublicationViewModelValidator()
	{
		RuleFor(model => model.Result).SetValidator(new PublicationValidator());
	}
}
```

Validator for Publication model:
```csharp

public sealed class PublicationValidator : AbstractValidator<Publication>
{
	public PublicationValidator()
	{
		RuleFor(model => model.Title).Length(4).WithMessage("Test message");
	}
}
```

In [example](https://github.com/canton7/Stylet/tree/master/Samples/Stylet.Samples.ModelValidation) with validation, the validated properties are directly inside the ViewModel, but if you put them in a separate class, the validation does not work. Is this a proper behavior?

**Comment by canton7** [comment, marked as answer]
It's to be expected. It's exactly the same problem as if you bind to properties of an object where that object does not implement INPC.

Validation works through `INotifyDataErrorInfo`. When you bind to a property, WPF checks to see whether the parent object implements that interface, and if it does, it subscribes to validation events.

The parent object of your `Title` property is the `Publication` class. So that's the class which needs to implement `INotifyDataErrorInfo` (usually, with Stylet, by subclassing `ValidatingModelBase`). 

This is just another application of the general WPF rule, which is "Do not bind to things which are not ViewModels"

**Reply by Catalan-1ap** [reply]
Thank you! 

Indeed, it is much easier to use individual properties rather than a separate class in a ViewModel. Implementation of INotifyDataErrorInfo as well as inheriting from ValidatingModelBase might spoil my Domain.

**Reply by canton7** [reply]
Yeah, separate your domain and presentation layers. Don't bind to domain objects (in all but the simplest cases): map them to presentation-layer ViewModels first
